mint-auditor/src/gnosis/fetcher.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Gnosis Safe transaction fetcher, used to get the transaction data from a
+//! gnosis safe-transaction-service.
+
+use super::{
+    api_data_types::{AllTransactionsResponse, RawGnosisTransaction},
+    Error, EthAddr,
+};
+use mc_common::logger::{log, o, Logger};
+use reqwest::{blocking::Client, StatusCode};
+use url::Url;
+
+/// Gnosis Safe transaction fetcher, used to get the transaction data from a
+/// gnosis safe-transaction-service.
+pub struct GnosisSafeFetcher {
+    /// Base URL for the gnosis safe-transaction-service API.
+    base_url: Url,
+
+    /// The [reqwest::Client].
+    client: Client,
+
+    /// Logger.
+    logger: Logger,
+}
+
+impl GnosisSafeFetcher {
+    /// Instantiate a [GnosisSafeFetcher] fetching transactions from the given
+    /// URL endpoint.
+    /// The URL endpoint is expected to run the Gnosis safe-transaction-service
+    /// (https://github.com/safe-global/safe-transaction-service/)
+    pub fn new(mut base_url: Url, logger: Logger) -> Result<Self, Error> {
+        if !base_url.path().ends_with('/') {
+            base_url = base_url.join(&format!("{}/", base_url.path()))?;
+        }
+
+        let logger = logger.new(o!("url" => base_url.to_string()));
+
+        let client = Client::builder()
+            .build()
+            .map_err(|e| Error::Other(format!("Failed to create reqwest client: {}", e)))?;
+
+        Ok(Self {
+            base_url,
+            client,
+            logger,
+        })
+    }
+
+    /// Fetch transaction data.
+    /// This method will continously reach out to the API service until all
+    /// transactions have been fetched. Because of pagination, this might result
+    /// in multiple requests. This returns only transactions that were
+    /// executed and confirmed.
+    pub fn get_all_transaction_data(
+        &self,
+        safe_address: &EthAddr,
+    ) -> Result<Vec<RawGnosisTransaction>, Error> {
+        let mut url = self.base_url.join(&format!(
+            "api/v1/safes/{}/all-transactions/?executed=true&queued=false&trusted=true",
```

Comment:
I think if we're calling mostly one endpoint this is fine. But if there are more coming, perhaps we want to refactor these up to an enum that stores the endpoint calls in one place?

Reply:
I don't currently see much value in it either unless we start hitting multiple gnosis endpoints.